folding/src/standard_config.rs

```diff
@@ -0,0 +1,471 @@
+            Side::Right => &self.witnesses[1],
+        };
+        &witness[*s]
+    }
+}
+
+/// contains a data structure useful to support the [CurrOrNext::Next] case
+/// in [FoldingEnv::col]
+mod memoization {
+    use ark_ff::Field;
+    use std::{
+        cell::{OnceCell, RefCell},
+        collections::HashMap,
+        hash::Hash,
+        sync::atomic::{AtomicUsize, Ordering},
+    };
+
+    /// a segment with up to N stored columns, and the potential
+    /// next segment, similar to a linked list N-length arrays
+    pub struct ColumnMemoizerSegment<F: Field, const N: usize> {
+        cols: [OnceCell<Vec<F>>; N],
+        next: OnceCell<Box<Self>>,
+    }
+
+    impl<F: Field, const N: usize> ColumnMemoizerSegment<F, N> {
+        pub fn new() -> Self {
+            let cols = [(); N].map(|_| OnceCell::new());
+            let next = OnceCell::new();
+            Self { cols, next }
+        }
+        // This will find the column if i < N, and get a reference to it,
+        // initializing it with `f` if needed.
+        // If i >= N it will continue recursing to the next segment, initializing
+        // it if needed
+        pub fn get_or_insert<I>(&self, i: usize, f: I) -> &Vec<F>
+        where
+            I: FnOnce() -> Vec<F>,
+        {
+            match i {
+                i if i < N => {
+                    let col = &self.cols[i];
+                    col.get_or_init(f)
+                }
+                i => {
+                    let i = i - N;
+                    let new = || Box::new(Self::new());
+                    let next = self.next.get_or_init(new);
+                    next.get_or_insert(i, f)
+                }
+            }
+        }
+    }
+    /// a hashmap like data structure supporting get-or-insert with
+    /// an immutable reference and returning an inmutable reference
+    /// without guard
+    pub struct ColumnMemoizer<C: Hash + Eq, F: Field, const N: usize> {
+        first_segment: ColumnMemoizerSegment<F, N>,
+        next: AtomicUsize,
+        ids: RefCell<HashMap<C, usize>>,
+    }
+
+    impl<C: Hash + Eq, F: Field, const N: usize> ColumnMemoizer<C, F, N> {
+        pub fn new() -> Self {
+            let first_segment = ColumnMemoizerSegment::new();
+            let next = AtomicUsize::from(0);
+            let ids = RefCell::new(HashMap::new());
+            Self {
+                first_segment,
+                next,
+                ids,
+            }
+        }
+        pub fn get_or_insert<I>(&self, col: C, f: I) -> &Vec<F>
+        where
+            I: FnOnce() -> Vec<F>,
+        {
+            // this will find or asign an id for the column and then
+            // search the segments using the id
+            let mut ids = self.ids.borrow_mut();
+            let new_id = || self.next.fetch_add(1, Ordering::Relaxed);
+            let id = ids.entry(col).or_insert_with(new_id);
+            self.first_segment.get_or_insert(*id, f)
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
I start having a preference to move tests into the subdirectory `tests`. I would move it in `tests/test_standard_config.rs`. But not blocking ofc.

Reply:
right, I actually wanted just `#[cfg(feature = "bn254")]`, as I don't have any proper test, just a function that shows the code builds, but I don't run it as that would creating the pairs and some prover/checker, which is already covered in the tests we have.

Reply:
I'll remove `#[cfg(test)]` for nwo